src/backend/oneapi/solve.cpp

```diff
@@ -204,56 +255,65 @@ Array<T> leastSquares(const Array<T> &a, const Array<T> &b) {
-        Array<T> tmp = createEmptyArray<T>(dim4(NUM));
-
-        vector<T> h_tau(NUM);
-
-        int info      = 0;
-        Buffer *A_buf = A.get();
-        Buffer *B_buf = B.get();
-        Buffer *dT    = tmp.get();
-
-        magma_geqrf3_gpu<T>(M, N, (*A_buf)(), A.getOffset(), A.strides()[1],
-                            &h_tau[0], (*dT)(), tmp.getOffset(), getQueue()(),
-                            &info);
-
-        int NRHS   = B.dims()[1];
-        int lhwork = (M - N + NB) * (NRHS + NB) + NRHS * NB;
-
-        vector<T> h_work(lhwork);
-        h_work[0] = scalar<T>(lhwork);
-
-        magma_unmqr_gpu<T>(MagmaLeft, MagmaConjTrans, M, NRHS, N, (*A_buf)(),
-                           A.getOffset(), A.strides()[1], &h_tau[0], (*B_buf)(),
-                           B.getOffset(), B.strides()[1], &h_work[0], lhwork,
-                           (*dT)(), tmp.getOffset(), NB, queue, &info);
-
-        magmablas_swapdblk<T>(MN - 1, NB, (*A_buf)(), A.getOffset(),
-                              A.strides()[1], 1, (*dT)(),
-                              tmp.getOffset() + NB * MN, NB, 0, queue);
-
-        if (getActivePlatform() == AFCL_PLATFORM_NVIDIA) {
-            Array<T> AT    = transpose<T>(A, true);
-            Buffer *AT_buf = AT.get();
-            OPENCL_BLAS_CHECK(gpu_blas_trsm(
-                OPENCL_BLAS_SIDE_LEFT, OPENCL_BLAS_TRIANGLE_LOWER,
-                OPENCL_BLAS_CONJ_TRANS, OPENCL_BLAS_NON_UNIT_DIAGONAL, N, NRHS,
-                scalar<T>(1), (*AT_buf)(), AT.getOffset(), AT.strides()[1],
-                (*B_buf)(), B.getOffset(), B.strides()[1], 1, &queue, 0,
-                nullptr, &event));
-        } else {
-            OPENCL_BLAS_CHECK(gpu_blas_trsm(
-                OPENCL_BLAS_SIDE_LEFT, OPENCL_BLAS_TRIANGLE_UPPER,
-                OPENCL_BLAS_NO_TRANS, OPENCL_BLAS_NON_UNIT_DIAGONAL, N, NRHS,
-                scalar<T>(1), (*A_buf)(), A.getOffset(), A.strides()[1],
-                (*B_buf)(), B.getOffset(), B.strides()[1], 1, &queue, 0,
-                nullptr, &event));
+        // Get workspace needed for QR
+        std::int64_t scratchpad_size =
+            ::oneapi::mkl::lapack::geqrf_scratchpad_size<compute_t<T>>(
+                getQueue(), M, N, A.strides()[1]);
+
+        sycl::buffer<compute_t<T>> scratchpad(scratchpad_size);
+        Array<compute_t<T>> t = createEmptyArray<T>(af::dim4(MN, 1, 1, 1));
+
+        sycl::buffer<compute_t<T>> aBuf =
+            A.template getBufferWithOffset<compute_t<T>>();
+        sycl::buffer<compute_t<T>> tBuf =
+            t.template getBufferWithOffset<compute_t<T>>();
+        // In place Perform in place QR
+        ::oneapi::mkl::lapack::geqrf(getQueue(), M, N, aBuf, A.strides()[1],
+                                     tBuf, scratchpad, scratchpad_size);
+
+        // matmul(Q1, B)
+        sycl::buffer<compute_t<T>> bBuf =
+            B.template getBufferWithOffset<compute_t<T>>();
+        if constexpr (std::is_same_v<compute_t<T>, float> ||
```

Comment:
we should create a variadic version of is_same_v that accepts any number of types. Reading this is hard on the eyes. something lke this:

is_any_of<T, double, float>

Reply:
had to bring out the candles and soft jazz for this one